### PR TITLE
Adds Veeps case

### PR DESCRIPTION
**Pull request by hugobarauna**

### CODE REVIEW

File: _posts/2024-03-05-veeps-elixir-case.markdown [hunk 1]
```diff
@@ -0,0 +1,134 @@
...
+
+Early on, the Veeps backend was implemented in Ruby on Rails. Its first version could handle a few thousand simultaneous users watching a concert without any impact to stream quality, which was fine when you have a handful of shows but would be insufficient with the expected show load and massive increase in concurrent viewership across streams. It was around that time that Vincent Franco joined Veeps as their CTO
+
+He had an extensive background in building and maintaining ticketing and event management software at scale. So he used that experience to further improve the system to handle tens of thousands of concurrent users. However, it became clear that improving it to hundreds of thousands would be a difficult challenge requiring substantial engineering efforts and increased operational costs, so the team began evaluating other stacks that could provide the out-of-the-box tooling for scaling in order to reach both short and long-term goals.
+
+
+## Adopting Elixir, hiring, and rewriting the system
+
+Vincent, who had successfully deployed Elixir as part of high-volume systems, believed Elixir was an excellent fit for Veeps' requirements.
```
Comment (whatyouhide) [inline _posts/2024-03-05-veeps-elixir-case.markdown:35]:
```suggestion
Vincent, who had successfully deployed Elixir as part of high-volume systems in the past, believed Elixir was an excellent fit for Veeps' requirements.
```

File: _posts/2024-03-05-veeps-elixir-case.markdown [hunk 2]
```diff
@@ -0,0 +1,134 @@
...
+
+Backed by his experience and several case studies from the Elixir community, such as [the one from Discord](/blog/2020/10/08/real-time-communication-at-scale-with-elixir-at-discord/), Vincent convinced management that Elixir could address their immediate scaling needs and become a reliable foundation on which the company could build.
+
+With buy-in from management, the plan was set in motion. They had two outstanding goals:
+
+* Prepare the platform to welcome the most famous artists in the world
+* Build their own team of engineers to help innovate and evolve the product
```
Comment (whatyouhide) [inline _posts/2024-03-05-veeps-elixir-case.markdown:42]: Periods

```suggestion
* Prepare the platform to welcome the most famous artists in the world.
* Build their own team of engineers to help innovate and evolve the product.
```

File: _posts/2024-03-05-veeps-elixir-case.markdown [hunk 3]
```diff
@@ -0,0 +1,134 @@
...
+
+Vincent knew that hiring right-fit technical people can take time and he didn't want to rush the process. Hence, he hired [DockYard](https://dockyard.com/) to rebuild the system while simultaneously searching for the right candidates to build out the team.
+
+Eight months later, the system had been entirely rewritten in Elixir and Phoenix. Phoenix Channels were used to enrich the live concert experience, while Phoenix LiveView empowered the ticket shopping journey.
+
+The rewrite was put to the test shortly after with a livestream that remains one of Veeps’ biggest, still to this day. Before the rewrite, 20 Rails nodes were used during big events, whereas now, only 2 Elixir nodes are required. The new platform was able to handle 83x more concurrent users than the previous system which transparently, was already ahead of competitors.
```
Comment (whatyouhide) [inline _posts/2024-03-05-veeps-elixir-case.markdown:48]:
```suggestion
The rewrite was put to the test shortly after with a livestream that remains one of Veeps’ biggest, still to this day. Before the rewrite, 20 Rails nodes were used during big events, whereas now, only 2 Elixir nodes were required. The new platform was able to handle 83x more concurrent users than the previous system which transparently, was already ahead of competitors.
```

Also, something got borked with this sentence:

> more concurrent users than the previous system which transparently, was already ahead of competitors

Can you fix it?

File: _posts/2024-03-05-veeps-elixir-case.markdown [hunk 4]
```diff
@@ -0,0 +1,134 @@
...
+
+The increase in infrastructure efficiency significantly reduced the need for complex auto-scaling solutions while providing ample capacity to handle high-traffic spikes.
+
+> The rewrite marked the most extensive and intricate system migration in my career, and yet, it was also the smoothest.
```
Comment (whatyouhide) [inline _posts/2024-03-05-veeps-elixir-case.markdown:52]: Man this quote is awesome

File: _posts/2024-03-05-veeps-elixir-case.markdown [hunk 5]
```diff
@@ -0,0 +1,134 @@
...
+
+
+## Perceived benefits of using Elixir and its ecosystem
+
+After using Elixir for more than two years, Veeps has experienced significant benefits. Here are a few of them.
+
+
+### Architectural simplicity
+
+Veeps' system has different scalability requirements for its various parts. For instance, when streaming a concert, the backend receives metadata from users' devices every 30 seconds to track viewership. This is the so-called Beaconing service.
```
Comment (whatyouhide) [inline _posts/2024-03-05-veeps-elixir-case.markdown:69]: Nits and rephrase

```suggestion
Different parts of the Veeps system have different scalability requirements. For instance, when streaming a show, the backend receives metadata from users' devices every 30 seconds to track viewership. This is the so-called *Beaconing service*.
```

File: _posts/2024-03-05-veeps-elixir-case.markdown [hunk 6]
```diff
@@ -0,0 +1,134 @@
...
+
+Say you have 250,000 people watching a concert — the Beaconing service handles thousands of requests per second for a few hours at a time. As a result, it needs to scale differently from other parts of the system, such as their merchandise e-commerce or backstage management.
```
Comment (whatyouhide) [inline _posts/2024-03-05-veeps-elixir-case.markdown:71]:
```suggestion
Say you have 250,000 people watching a concert: the Beaconing service needs to handle thousands of requests per second for a few hours at a time. As a result, it needs to scale differently from other parts of the system, such as the merchandise e-commerce or backstage management.
```

File: _posts/2024-03-05-veeps-elixir-case.markdown [hunk 7]
```diff
@@ -0,0 +1,134 @@
...
+
+To tackle this issue, they built a distributed system. They packaged each subsystem as an [Elixir release](https://hexdocs.pm/elixir/config-and-releases.html#releases), totaling five releases. For the communication layer, they used distributed Erlang, which is built into Erlang/OTP, allowing seamless inter-process communication across networked nodes.
+
+In a nutshell, each node contains several processes with specific responsibilities. Each of these processes belongs to their respective [distributed process group](https://www.erlang.org/doc/man/pg.html). If node A needs billing information, it will reach out to any process within the "billing process group", which may be anywhere in the cluster.
+
+When deploying a new version of the system, they deploy a new cluster altogether, with all five subsystems at once. Given Elixir's scalability, the whole system uses 9 nodes, making a simple deployment strategy affordable and practical. As we will see, this approach is well-supported during development too, thanks to the use of Umbrella Projects.
+
+
+### Service-oriented architecture within a Monorepo
```
Comment (whatyouhide) [inline _posts/2024-03-05-veeps-elixir-case.markdown:80]: Nit: I don't think we need to upcase "Monorepo", it's not a proper noun and other headers are not upcased. We prolly shouldn't upcase "Monorepo" in the rest of this section too

```suggestion
### Service-oriented architecture within a monorepo
```

File: _posts/2024-03-05-veeps-elixir-case.markdown [hunk 8]
```diff
@@ -0,0 +1,134 @@
...
+
+Although they run a distributed system, they organize the code in only one repository, following the Monorepo approach. To do that, they use the [Umbrella Project feature](https://hexdocs.pm/elixir/dependencies-and-umbrella-projects.html#content) from Mix, a build tool that ships with Elixir.
```
Comment (whatyouhide) [inline _posts/2024-03-05-veeps-elixir-case.markdown:82]: We don't have that many build tools that ship with Elixir 😄 

```suggestion
Although they run a distributed system, they organize the code in only one repository, following the Monorepo approach. To do that, they use the [Umbrella Project feature](https://hexdocs.pm/elixir/dependencies-and-umbrella-projects.html#content) from Mix, th build tool that ships with Elixir.
```

File: _posts/2024-03-05-veeps-elixir-case.markdown [hunk 9]
```diff
@@ -0,0 +1,134 @@
...
+
+Their umbrella project consists of 16 applications (at the time of writing), which they [sliced into five OTP releases](https://hexdocs.pm/mix/Mix.Tasks.Release.html#module-umbrellas). The remaining applications contain code that needs to be shared between multiple applications. For example, one of the shared applications defines all structs sent as messages across the subsystems, guaranteeing all subsystems have a consistent view of the data.
+
+> With the Umbrella Project, you can have the developer experience benefits of a single code repository while having a service-oriented architecture.
+>
+> \- Andrea Leopardi, Principal Engineer
```
Comment (whatyouhide) [inline _posts/2024-03-05-veeps-elixir-case.markdown:88]:
```suggestion
> With umbrella projects, you can have the developer experience benefits of a single code repository, while being able to build a service-oriented architecture.
>
> \- Andrea Leopardi, Principal Engineer
```

File: _posts/2024-03-05-veeps-elixir-case.markdown [hunk 10]
```diff
@@ -0,0 +1,134 @@
...
+
+When deploying a new version of the system, they deploy a new cluster altogether, with all five subsystems at once. Given Elixir's scalability, the whole system uses 9 nodes, making a simple deployment strategy affordable and practical. As we will see, this approach is well-supported during development too, thanks to the use of Umbrella Projects.
+
+
+### Service-oriented architecture within a Monorepo
+
+Although they run a distributed system, they organize the code in only one repository, following the Monorepo approach. To do that, they use the [Umbrella Project feature](https://hexdocs.pm/elixir/dependencies-and-umbrella-projects.html#content) from Mix, a build tool that ships with Elixir.
+
+Their umbrella project consists of 16 applications (at the time of writing), which they [sliced into five OTP releases](https://hexdocs.pm/mix/Mix.Tasks.Release.html#module-umbrellas). The remaining applications contain code that needs to be shared between multiple applications. For example, one of the shared applications defines all structs sent as messages across the subsystems, guaranteeing all subsystems have a consistent view of the data.
```
Comment (whatyouhide) [inline _posts/2024-03-05-veeps-elixir-case.markdown:84]: Nit: the structs don't techincally guarantee a consistent view of the data, only shared schemas.

```suggestion
Their umbrella project consists of 16 applications (at the time of writing), which they [sliced into five OTP releases](https://hexdocs.pm/mix/Mix.Tasks.Release.html#module-umbrellas). The remaining applications contain code that needs to be shared between multiple applications. For example, one of the shared applications defines all the structs sent as messages across the subsystems, guaranteeing that all subsystems use the same schemas for that exchanged data.
```

File: _posts/2024-03-05-veeps-elixir-case.markdown [hunk 11]
```diff
@@ -0,0 +1,134 @@
...
+
+For caching, the team relies on [ETS (Erlang Term Storage)](https://www.erlang.org/doc/man/ets.html), a high-performing key-value store part of the Erlang standard library. For cache busting between multiple parts of the distributed system, they use [Phoenix PubSub](https://github.com/phoenixframework/phoenix_pubsub), a real-time publisher/subscriber library that comes with [Phoenix](https://phoenixframework.org/).
+
+Before the rewrite, the Beaconing service used Google's Firebase. Now, the system uses [Broadway](https://elixir-broadway.org/) to ingest data from hundreds of thousands of concurrent users. Broadway is an Elixir library for building concurrent data ingestion and processing pipelines. They utilized the library's capabilities to efficiently send requests to AWS services, regulating batch sizes to comply with AWS limits. They also used it to handle rate limiting to adhere to AWS service constraints. All of this was achieved with Broadway's built-in functionality.
```
Comment (whatyouhide) [inline _posts/2024-03-05-veeps-elixir-case.markdown:99]:
```suggestion
Before the rewrite, the Beaconing service used Google's Firebase. Now, the system uses [Broadway](https://elixir-broadway.org/) to ingest data from hundreds of thousands of HTTP requests from concurrent users. Broadway is an Elixir library for building concurrent data ingestion and processing pipelines. They utilized the library's capabilities to efficiently send requests to AWS services, regulating batch sizes to comply with AWS limits. They also used it to handle rate limiting to adhere to AWS service constraints. All of this was achieved with Broadway's built-in functionality.
```

File: _posts/2024-03-05-veeps-elixir-case.markdown [hunk 12]
```diff
@@ -0,0 +1,134 @@
...
+
+Finally, they use [Oban](https://getoban.pro/), an Elixir library for background jobs.
```
Comment (whatyouhide) [inline _posts/2024-03-05-veeps-elixir-case.markdown:101]:
```suggestion
Finally, they use [Oban](https://getoban.pro/), an Elixir library for background jobs, for all sorts of background-work use cases.
```

File: _posts/2024-03-05-veeps-elixir-case.markdown [hunk 13]
```diff
@@ -0,0 +1,134 @@
...
+
+Throughout the development journey, Veeps consistently found that Elixir and its ecosystem had built-in solutions for their technical challenges. Here's what Vincent, CTO of Veeps, had to say about that:
+
+> Throughout my career, I've worked with large-scale systems at several companies. However, at Veeps, it's unique because we achieve this scale with minimal reliance on external tools. It's primarily just Elixir and its ecosystem that empower us.
+>
+> \- Vincent Franco, CTO
+
+This operational simplicity benefited not only the production environment but also the development side. The team could focus on learning Elixir and its ecosystem without the need to master additional technologies, resulting in increased productivity.
```
Comment (whatyouhide) [inline _posts/2024-03-05-veeps-elixir-case.markdown:109]: Typo.

```suggestion
This operational simplicity benefitted not only the production environment but also the development side. The team could focus on learning Elixir and its ecosystem without the need to master additional technologies, resulting in increased productivity.
```

One more thing: you casually mention that the team could focus on **learning** Elixir, but you never mentioned that the team was new to Elixir. Maybe something like this:

```suggestion
This operational simplicity benefited not only the production environment but also the development side. The team, in which most engineers had no Elixir experience before Veeps, could focus on learning Elixir and its ecosystem without the need to master additional technologies, resulting in increased productivity.
```

File: _posts/2024-03-05-veeps-elixir-case.markdown [hunk 14]
```diff
@@ -0,0 +1,134 @@
...
+
+
+### LiveView: simplifying the interaction between front-end and back-end developers
+
+After the rewrite, [LiveView](https://github.com/phoenixframework/phoenix_live_view), a Phoenix library for building interactive, real-time web apps, was used for every part of the front-end except for the Onstage subsystem.
```
Comment (whatyouhide) [inline _posts/2024-03-05-veeps-elixir-case.markdown:114]:
```suggestion
After the rewrite, [LiveView](https://github.com/phoenixframework/phoenix_live_view), a Phoenix library for building interactive, real-time web apps, was used for every part of the front-end except for the "Onstage" subsystem (responsible for the live stream itself).
```

File: _posts/2024-03-05-veeps-elixir-case.markdown [hunk 15]
```diff
@@ -0,0 +1,134 @@
...
+
+The two front-end developers, who came from a React background, also started writing LiveView. After this new experience, the team found the process of API negotiation between the front-end and back-end engineers much simpler compared to when using React. This was because they only had to use Elixir modules and functions instead of creating additional HTTP API endpoints and all the extra work that comes with them, such as API versioning.
+
+> Our front-end team, originally proficient in React, has made a remarkable transition to LiveView. They've wholeheartedly embraced its user-friendly nature and its smooth integration into our system.
+>
+> \- Vincent Franco, CTO
+
+
+## Conclusion: insights from Veeps' Elixir Experience
```
Comment (whatyouhide) [inline _posts/2024-03-05-veeps-elixir-case.markdown:123]: Mixed casing here

```suggestion
## Conclusion: insights from Veeps' Elixir experience
```

**Review by whatyouhide — COMMENTED** [review]
Left a bunch of minor comments

File: _posts/2024-03-05-veeps-elixir-case.markdown (hunk 3 above)
Comment (hugobarauna) [reply _posts/2024-03-05-veeps-elixir-case.markdown]: Done on https://github.com/elixir-lang/elixir-lang.github.com/pull/1753/commits/975f9e7d9f4eb8acc5ac9e59e794e1ddf866402b

**Comment by hugobarauna** [comment]
Thanks for the awesome review, @whatyouhide !  ❤️ 🧡 💛 💚 💙 💜 🤎 🖤 🤍

**Review by whatyouhide — APPROVED** [review]
:shipit:

**Comment by josevalim** [comment]
:green_heart: :blue_heart: :purple_heart: :yellow_heart: :heart:

**Pull request merged**